Fail to build

Using cargo 1.90.0-nightly (409fed7dc 2025-06-23) + LLVM 20.1.7:

```
error[E0308]: mismatched types
   --> src/components/app.rs:561:62
    |
561 |             subscriptions.push(settings_daemon::subscription(connection).map(Msg::SettingsDaemon));
    |                                ----------------------------- ^^^^^^^^^^ expected `zbus::connection::Connection`, found `zbus::Connection`
    |                                |
    |                                arguments to this function are incorrect
    |
note: two different versions of crate `zbus` are being used; two types coming from two different versions of the same crate are different types even if they look the same
   --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/zbus-4.4.0/src/connection/mod.rs:193:1
    |
193 | pub struct Connection {
    | ^^^^^^^^^^^^^^^^^^^^^ this is the found type `zbus::Connection`
    |
   ::: /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/zbus-5.7.1/src/connection/mod.rs:208:1
    |
208 | pub struct Connection {
    | ^^^^^^^^^^^^^^^^^^^^^ this is the expected type `zbus::connection::Connection`
    |
   ::: src/components/app.rs:10:5
    |
10  | use cosmic::{
    |     ------ one version of crate `zbus` used here, as a dependency of crate `cosmic`
...
45  | use zbus::Connection;
    |     ---- one version of crate `zbus` used here, as a direct dependency of the current crate
    = help: you can use `cargo tree` to explore your dependency tree
note: function defined here
   --> /root/.cargo/git/checkouts/cosmic-settings-subscriptions-3e3ddc5284b25636/84ab9e8/src/settings_daemon.rs:11:8
    |
11  | pub fn subscription(connection: zbus::Connection) -> iced_futures::Subscription<Event> {
    |        ^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
warning: `cosmic-osd` (bin "cosmic-osd") generated 1 warning
error: could not compile `cosmic-osd` (bin "cosmic-osd") due to 1 previous error; 1 warning emitted
make: *** [Makefile:33: cosmic-osd] Error 101
```

No issues building other COSMIC projects.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -23,7 +23,7 @@ libcosmic = { git = "https://github.com/pop-os/libcosmic", default-features = fa
 ] }
 libpulse-binding = "2.28.1"
 log = "0.4.21"
-logind-zbus = "4.0.3"
+logind-zbus = "5.3.2"
 once_cell = "1.18.0"
 rust-embed = "8.0.0"
 rustix = "0.38.31"
@@ -32,7 +32,7 @@ serde_json = "1.0.139"
 tokio = { version = "1.23.0", features = ["process"] }
 tokio-stream = "0.1.11"
 uzers = "0.11.3"
-zbus = { version = "4.2.2", default-features = false, features = ["tokio"] }
+zbus = { version = "5.7.1", default-features = false, features = ["tokio"] }
 cosmic-settings-subscriptions = { git = "https://github.com/pop-os/cosmic-settings-subscriptions" }
 #[patch.'https://github.com/pop-os/libcosmic']
 #libcosmic = { path = "../../pop/libcosmic" }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,18 +1,18 @@
 [package]
 name = "cosmic-osd"
 version = "0.1.0"
-edition = "2021"
+edition = "2024"
 license = "GPL-3.0-or-later"
 
 [dependencies]
-clap = { version = "4.4.8", features = ["derive"] }
-env_logger = "0.11.3"
-futures = "0.3.30"
-i18n-embed = { version = "0.14", features = [
+clap = { version = "4.5.40", features = ["derive"] }
+env_logger = "0.11.8"
+futures = "0.3.31"
+i18n-embed = { version = "0.15", features = [
     "fluent-system",
     "desktop-requester",
 ] }
-i18n-embed-fl = "0.8"
+i18n-embed-fl = "0.9"
 libcosmic = { git = "https://github.com/pop-os/libcosmic", default-features = false, features = [
     "autosize",
     "tokio",
@@ -21,17 +21,15 @@ libcosmic = { git = "https://github.com/pop-os/libcosmic", default-features = fa
     "winit",
     "single-instance",
 ] }
-libpulse-binding = "2.28.1"
-log = "0.4.21"
+libpulse-binding = "2.30.1"
+log = "0.4.27"
 logind-zbus = "5.3.2"
-once_cell = "1.18.0"
-rust-embed = "8.0.0"
-rustix = "0.38.31"
-serde = { version = "1.0.140", features = ["derive"] }
-serde_json = "1.0.139"
-tokio = { version = "1.23.0", features = ["process"] }
-tokio-stream = "0.1.11"
-uzers = "0.11.3"
+rust-embed = "8.7.2"
+serde = { version = "1.0.219", features = ["derive"] }
+serde_json = "1.0.140"
+tokio = { version = "1.46.1", features = ["process"] }
+tokio-stream = "0.1.17"
+uzers = "0.12.1"
 zbus = { version = "5.7.1", default-features = false, features = ["tokio"] }
 cosmic-settings-subscriptions = { git = "https://github.com/pop-os/cosmic-settings-subscriptions" }
 #[patch.'https://github.com/pop-os/libcosmic']

--- a/src/components/app.rs
+++ b/src/components/app.rs
@@ -27,7 +27,7 @@ use cosmic::{
     },
     theme,
     widget::{autosize::autosize, button, container, icon, text, Column},
-    DbusActivation, Element,
+    Element,
 };
 use cosmic_settings_subscriptions::{
     airplane_mode, pulse, settings_daemon,

--- a/src/components/app.rs
+++ b/src/components/app.rs
@@ -8,30 +8,29 @@ use crate::{
 use crate::{cosmic_session::CosmicSessionProxy, session_manager::SessionManagerProxy};
 use clap::Parser;
 use cosmic::{
+    Element,
     app::{self, CosmicFlags, Task},
     dbus_activation::Details,
     iced::{
-        self,
+        self, Alignment, Length, Limits, Point, Rectangle, Size, Subscription,
         event::{
             self, listen_with,
             wayland::{self, LayerEvent, OverlapNotifyEvent},
         },
-        keyboard::{key::Named, Key},
+        keyboard::{Key, key::Named},
         time,
         window::Id as SurfaceId,
-        Alignment, Length, Limits, Point, Rectangle, Size, Subscription,
     },
     iced_runtime::platform_specific::wayland::layer_surface::SctkLayerSurfaceSettings,
     iced_winit::commands::layer_surface::{
-        destroy_layer_surface, get_layer_surface, Anchor, KeyboardInteractivity,
+        Anchor, KeyboardInteractivity, destroy_layer_surface, get_layer_surface,
     },
     theme,
-    widget::{autosize::autosize, button, container, icon, text, Column},
-    Element,
+    widget::{Column, autosize::autosize, button, container, icon, text},
 };
 use cosmic_settings_subscriptions::{
     airplane_mode, pulse, settings_daemon,
-    upower::kbdbacklight::{kbd_backlight_subscription, KeyboardBacklightUpdate},
+    upower::kbdbacklight::{KeyboardBacklightUpdate, kbd_backlight_subscription},
 };
 use logind_zbus::manager::ManagerProxy;
 use serde::{Deserialize, Serialize};
@@ -114,7 +113,7 @@ async fn log_out() -> zbus::Result<()> {
 }
 
 impl Display for OsdTask {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
         write!(f, "{}", serde_json::ser::to_string(self).unwrap())
     }
 }
@@ -182,7 +181,7 @@ struct App {
 
 impl App {
     fn create_indicator(&mut self, params: osd_indicator::Params) -> cosmic::app::Task<Msg> {
-        if let Some((_id, ref mut state)) = &mut self.indicator {
+        if let Some((_id, state)) = &mut self.indicator {
             state.replace_params(params)
         } else {
             let id = SurfaceId::unique();
@@ -605,7 +604,7 @@ impl cosmic::Application for App {
         unreachable!()
     }
 
-    fn view_window(&self, id: SurfaceId) -> cosmic::Element<'_, Msg> {
+    fn view_window(&self, id: SurfaceId) -> cosmic::Element<Msg> {
         if let Some(surface) = self.surfaces.get(&id) {
             return match surface {
                 Surface::PolkitDialog(state) => {
@@ -750,11 +749,11 @@ mod pipewire {
     }
 }
 
-fn min_width_and_height<'a>(
-    e: Element<'a, Msg>,
+fn min_width_and_height(
+    e: Element<Msg>,
     width: impl Into<Length>,
     height: impl Into<Length>,
-) -> Column<'a, Msg> {
+) -> Column<Msg> {
     use iced::widget::{column, horizontal_space, row, vertical_space};
     column![
         row![e, vertical_space().height(height)].align_y(Alignment::Center),

--- a/src/components/osd_indicator.rs
+++ b/src/components/osd_indicator.rs
@@ -2,23 +2,23 @@
 // TODO: Dismiss on click?
 
 use cosmic::{
-    iced::{self, window::Id as SurfaceId, Alignment, Border, Length},
+    Element, Task,
+    iced::{self, Alignment, Border, Length, window::Id as SurfaceId},
     iced_runtime::platform_specific::wayland::layer_surface::SctkLayerSurfaceSettings,
     iced_winit::commands::{
         layer_surface::{
-            destroy_layer_surface, get_layer_surface, Anchor, KeyboardInteractivity, Layer,
+            Anchor, KeyboardInteractivity, Layer, destroy_layer_surface, get_layer_surface,
         },
         overlap_notify::overlap_notify,
     },
     widget::{self, horizontal_space, vertical_space},
-    Element, Task,
 };
-use futures::future::{abortable, AbortHandle, Aborted};
-use once_cell::sync::Lazy;
+use futures::future::{AbortHandle, Aborted, abortable};
+use std::sync::LazyLock;
 use std::time::Duration;
 
-pub static OSD_INDICATOR_ID: Lazy<widget::Id> =
-    Lazy::new(|| widget::Id::new("osd-indicator".to_string()));
+pub static OSD_INDICATOR_ID: LazyLock<widget::Id> =
+    LazyLock::new(|| widget::Id::new("osd-indicator".to_string()));
 
 #[derive(Debug)]
 pub enum Params {
@@ -144,7 +144,7 @@ impl State {
         cmd
     }
 
-    pub fn view(&self) -> Element<'_, Msg> {
+    pub fn view(&self) -> Element<Msg> {
         let icon = widget::icon::from_name(self.params.icon_name());
 
         // Use large radius on value-OSD to enforce pill-shape with "Round" system style

--- a/src/components/polkit_dialog.rs
+++ b/src/components/polkit_dialog.rs
@@ -8,26 +8,24 @@ use crate::{
 };
 use cosmic::{
     iced::{
-        self,
-        event::{wayland, PlatformSpecific},
+        self, Subscription, Task,
+        event::{PlatformSpecific, wayland},
         window::Id as SurfaceId,
-        Subscription, Task,
     },
     iced_runtime::platform_specific::wayland::layer_surface::SctkLayerSurfaceSettings,
     iced_winit::wayland::commands::layer_surface::{
-        destroy_layer_surface, get_layer_surface, KeyboardInteractivity, Layer,
+        KeyboardInteractivity, Layer, destroy_layer_surface, get_layer_surface,
     },
     widget,
 };
-use once_cell::sync::Lazy;
 use std::{
     collections::HashMap,
-    sync::{Arc, Mutex},
+    sync::{Arc, LazyLock, Mutex},
 };
 use tokio::sync::oneshot;
 
-pub static POLKIT_DIALOG_ID: Lazy<widget::Id> =
-    Lazy::new(|| widget::Id::new("polkit-dialog".to_string()));
+pub static POLKIT_DIALOG_ID: LazyLock<widget::Id> =
+    LazyLock::new(|| widget::Id::new("polkit-dialog".to_string()));
 
 #[derive(Clone, Debug)]
 pub struct Params {
@@ -171,7 +169,7 @@ impl State {
         (Some(self), Task::none())
     }
 
-    pub fn view(&self) -> cosmic::Element<'_, Msg> {
+    pub fn view(&self) -> cosmic::Element<Msg> {
         // TODO Allocates on every keypress?
 
         let placeholder = self.password_label.trim_end_matches(':');

--- a/src/cosmic_session.rs
+++ b/src/cosmic_session.rs
@@ -8,6 +8,6 @@ use zbus::proxy;
     default_service = "com.system76.CosmicSession",
     default_path = "/com/system76/CosmicSession"
 )]
-trait CosmicSession {
+pub trait CosmicSession {
     fn exit(&self) -> zbus::Result<()>;
 }

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,16 +1,17 @@
 use i18n_embed::{
-    fluent::{fluent_language_loader, FluentLanguageLoader},
     DesktopLanguageRequester,
+    fluent::{FluentLanguageLoader, fluent_language_loader},
 };
-use once_cell::sync::Lazy;
 use rust_embed::RustEmbed;
+use std::sync::LazyLock;
 
 mod components;
 pub mod cosmic_session;
 pub mod session_manager;
 mod subscriptions;
 
-pub static LANG_LOADER: Lazy<FluentLanguageLoader> = Lazy::new(|| fluent_language_loader!());
+pub static LANG_LOADER: LazyLock<FluentLanguageLoader> =
+    LazyLock::new(|| fluent_language_loader!());
 
 #[derive(RustEmbed)]
 #[folder = "i18n"]

--- a/src/session_manager.rs
+++ b/src/session_manager.rs
@@ -24,7 +24,7 @@
 use zbus::proxy;
 
 #[proxy(interface = "org.gnome.SessionManager", assume_defaults = true)]
-trait SessionManager {
+pub trait SessionManager {
     /// CanRebootToFirmwareSetup method
     fn can_reboot_to_firmware_setup(&self) -> zbus::Result<bool>;
 

--- a/src/subscriptions/dbus.rs
+++ b/src/subscriptions/dbus.rs
@@ -46,14 +46,14 @@ pub fn subscription() -> iced::Subscription<Event> {
 }
 
 async fn connection() -> zbus::Result<zbus::Connection> {
-    zbus::ConnectionBuilder::session()?
+    zbus::connection::Builder::session()?
         .name(NAME)?
         .build()
         .await
 }
 
 async fn system_connection() -> zbus::Result<zbus::Connection> {
-    zbus::ConnectionBuilder::system()?.build().await
+    zbus::connection::Builder::system()?.build().await
 }
 
 fn result_to_event<T>(res: zbus::Result<T>, context: &'static str, f: fn(T) -> Event) -> Event {

--- a/src/subscriptions/polkit_agent.rs
+++ b/src/subscriptions/polkit_agent.rs
@@ -52,9 +52,8 @@ pub struct Subject<'a> {
 }
 
 impl<'a> zvariant::Type for Subject<'a> {
-    fn signature() -> zvariant::Signature<'static> {
-        unsafe { zvariant::Signature::from_bytes_unchecked(b"(sa{sv})") }
-    }
+    const SIGNATURE: &'static zvariant::Signature =
+        <(String, HashMap<String, zvariant::Value>)>::SIGNATURE;
 }
 
 #[derive(serde::Serialize, serde::Deserialize)]
@@ -64,9 +63,8 @@ pub struct Identity<'a> {
 }
 
 impl<'a> zvariant::Type for Identity<'a> {
-    fn signature() -> zvariant::Signature<'static> {
-        unsafe { zvariant::Signature::from_bytes_unchecked(b"(sa{sv})") }
-    }
+    const SIGNATURE: &'static zvariant::Signature =
+        <(String, HashMap<String, zvariant::Value>)>::SIGNATURE;
 }
 
 #[zbus::proxy(

--- a/src/subscriptions/polkit_agent_helper.rs
+++ b/src/subscriptions/polkit_agent_helper.rs
@@ -117,7 +117,7 @@ impl Responder {
 }
 
 impl fmt::Debug for Responder {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "Responder")
     }
 }

